### ISSUE TITLE
Rust's `i128` return ABI does not agree with Clang and GCC on Windows targets

### ISSUE THREAD

**Issue by tgross35** [rust-lang/rust#134288]
Originally reported at https://github.com/rust-lang/lang-team/issues/255#issuecomment-2214459945, Clang and GCC-MinGW return `__int128` in xmm0. Rust currently returns it on the stack. 

@wesleywiser has a proposed patch to make Clang return on the stack for `-msvc` targets, while keeping the behavior for MinGW https://github.com/wesleywiser/llvm-project/commit/b2f8b8308bc25f81debe5bd99b02f497fdcc372a.

Rustc could probably just change the Windows ABI to always return in xmm0 for now, and then if the Clang patch lands we can change this to apply only to MinGW.

Demo: https://godbolt.org/z/o9h8We69o

**Comment by joshtriplett** [comment]
What are the chances that we can get mingw to fix their alignment to match MSVC? The two are *supposed* to be interoperable.

**Comment by tgross35** [comment]
I think that the Windows story might actually be all set since https://github.com/rust-lang/rust/pull/134290 merged; the alignment is the same (looking at https://godbolt.org/z/WeEsYTqfP), and MinGW, Clang, and (as of a few minutes ago) Rust all pass `i128` on the stack but return in xmm0.

I left this open because @wesleywiser may still plan to submit Clang patch, which is linked in the top post. But maybe that is no longer needed? Unless Microsoft eventually specifies an `__int128` ABI, at which point all implementations will need to update. (Or they can just bless passing on the stack and returning in XMM, in which case nobody has to change anything :) )

**Comment by mati865** [comment]
Correct, until MS implements i128 the only way to be compatible with them is to remove i128 on Windows.

**Comment by RalfJung** [comment]
We can do whatever we want on our own ABI, but we'd have to remove it from `extern "C"` functions.

**Comment by RalfJung** [comment]
I was just replying to @mati865. No strong opinion from my side; I just think there should be an explicit decision by a team (compiler and/or lang) on our stance of whether we can promise to be ABI-compatible with the native C calling convention for a type that the "official" C compiler for that target *does not support*.
